File: StriverA2Z/JavaScript&TypeScript/src/Arrays/Medium/sortZeroOneTwo.ts
function zeroOneTwo(arr: number[]) {
  let low = 0,
    mid = 0,
    high = arr.length - 1;
  while (mid <= high) {
    if (arr[mid] === 0) {
      //swap
      [arr[low], arr[mid]] = [arr[mid], arr[low]];
      low++;
      mid++;
    } else if (arr[mid] === 1) {
      mid++;
    } else {
      //swap
      [arr[mid], arr[high]] = [arr[high], arr[mid]];
      high--;
    }
  }
  console.log(arr);
}

zeroOneTwo([0, 1, 2, 1, 0, 1, 2, 1, 0, 0, 0, 2, 2, 2, 1, 1, 0]);

File: StriverA2Z/C++/3_ArraysQuestion/3.2_Medium/sort_zero_one_two.cpp
#include<bits/stdc++.h>

using namespace std;

void swap(int arr[], int start, int end) {
    int temp = arr[start];
    arr[start] = arr[end];
    arr[end] = temp;
}

void dutchNationalFlagAlgorithm(int arr[], int n) {
    int low = 0;
    int mid = 0;
    int high = n - 1;
    while (mid <= high) {
        if (arr[mid] == 0) {
            swap(arr, low, mid);
            low++;
            mid++;
        } else if (arr[mid] == 1) {
            mid++;
        } else {
            swap(arr, mid, high);
            high--;
        }
    }

}

int main() {
    int arr[] = {1, 0, 0, 0, 1, 1, 1, 2, 2, 0, 2, 1, 0};
    dutchNationalFlagAlgorithm(arr, sizeof(arr) / sizeof(arr[0]));
    for (int x: arr) cout << x << " ";
}